devito/types/basic.py

```diff
@@ -464,50 +462,47 @@ class AbstractFunction(sympy.Function, Basic, Pickable):
 
     is_AbstractFunction = True
 
-
-class AbstractCachedFunction(AbstractFunction, Cached, Evaluable):
-    """
-    Base class for tensor symbols, cached by both Devito and Sympy.
-
-    For more information, refer to ``AbstractFunction.__doc__``.
-    """
-
     def __new__(cls, *args, **kwargs):
         options = kwargs.get('options', {})
-        if cls in _SymbolCache:
+        if cls._cached():
             newobj = sympy.Function.__new__(cls, *args, **options)
             newobj._cached_init()
         else:
             name = kwargs.get('name')
             indices = cls.__indices_setup__(**kwargs)
 
+            # Create new, unique type instance from cls and the symbol name
+            newcls = type(name, (cls,), dict(cls.__dict__))
+
             # Create the new Function object and invoke __init__
-            newcls = cls._symbol_type(name)
             newobj = sympy.Function.__new__(newcls, *indices, **options)
 
             # Initialization. The following attributes must be available
-            # when executing __init__
+            # when executing __init_finalize__
             newobj._name = name
             newobj._indices = indices
             newobj._shape = cls.__shape_setup__(**kwargs)
             newobj._dtype = cls.__dtype_setup__(**kwargs)
-            newobj.__init__(*args, **kwargs)
+            newobj.__init_finalize__(*args, **kwargs)
 
-            # All objects cached on the AbstractFunction /newobj/ keep a reference
-            # to /newobj/ through the /function/ field. Thus, all indexified
-            # object will point to /newobj/, the "actual Function".
+            # All objects cached on the AbstractFunction `newobj` keep a reference
+            # to `newobj` through the `function` field. Thus, all indexified
+            # object will point to `newobj`, the "actual Function".
             newobj.function = newobj
 
             # Store new instance in symbol cache
-            newcls._cache_put(newobj)
+            Cached.__init__(newobj, newcls)
         return newobj
 
     def __init__(self, *args, **kwargs):
-        if not self._cached():
```

Comment:
note for reviewers: we don't need `if not self._cached()` anymore. Now we just do the real init in ``_init_finalize__``